### PR TITLE
[rfw] Remote Flutter Widgets package

### PR DESCRIPTION
**Pull request by Hixie**
This package defines a format for describing Flutter widget trees in
text or binary files that can be loaded dynamically at runtime.

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by Hixie** [comment]
For review purposes I have split this into two commits, one containing mostly only boilerplate from `flutter create` and mechanical changes like adding license blocks, and the other containing files that are potentially interesting to review.

**Comment by stuartmorgan** [comment]
For anything you want to temporarily exclude on stable, in .cirrus.yml replace:
```
- ./script/tool_runner.sh this-will-fail-on-stable ...
```

with:

```
# Exclude rfw until <insert version here> because <reason>
- if [[ "$CHANNEL" == "master" ]]; then
-   ./script/tool_runner.sh this-will-fail-on-stable ...
- else
-   ./script/tool_runner.sh this-will-fail-on-stable ... --exclude=rfw
- fi
```

(We don't need to do this often enough to have built this ability into the tool; if it becomes an issue we certainly could.)

**Comment by chgibb** [comment]
What are the intended use cases for this package?

**Comment by Hixie** [comment]
@chgibb does the README in the PR provide any illumination? https://github.com/flutter/packages/blob/8cb0765ae57775abfc3fa167e29fc2bd33ff544c/packages/rfw/README.md

**Comment by chgibb** [comment]
Thanks for pointing that out! Indeed it does.

It looks like the `rfw` language is very focused on declarative UI and is not Turing-complete. Is this an explicit design goal? Is the focus on server-driven UI / UI as configuration?

**Comment by Hixie** [comment]
Yes, this package's goal is just declarative UI. Imperative scripting is a problem already solved by other technologies, such as JavaScript, Wasm, or Lua. Or Dart, in principle, though in practice today we don't have a good way to execute arbitrary Dart dynamically at runtime on phones.

**Comment by chgibb** [comment]
Enabling this kind of out of band functionality is something I am enthralled with. If you can point me to design documents, blogs or other content digging into the motivations for a first-party package like `rfw`, I would be incredibly grateful.

**Comment by Hixie** [comment]
@chgibb All the documentation is in the PR itself (there's extensive dartdocs, the README, and four examples). This was just an experiment, there wasn't really a motivating factor beyond my playing with what we could do in the space of server-driven UIs, which is something people ask for occasionally (see the discussion of use cases in the README in the PR). If you have any specific questions I'm more than happy to answer them. :-)

**Comment by Hixie** [comment]
I've added a commit which addresses the review comments found in:
https://github.com/Hixie/packages/commit/8cb0765ae57775abfc3fa167e29fc2bd33ff544c

**Review by dnfield — APPROVED** [review]
LGTM. None of my outstanding comments are blocking.

**Comment by dnfield** [comment]
(I must admit I have not carefully reviewed every line of code, but I did try to look more at docs and surface API, as well as tests, which look good enough to get this going to me)

File: packages/rfw/example/wasm/lib/main.dart [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    super.initState();
+    RendererBinding.instance!.deferFirstFrame();
+    _runtime.update(const LibraryName(<String>['core', 'widgets']), createCoreWidgets());
+    _loadLogic();
+  }
+
+  late final WasmFunction _dataFetcher;
+
+  Future<void> _loadLogic() async {
+    final DateTime expiryDate = DateTime.now().subtract(const Duration(hours: 6));
```
Comment (treeplate) [inline packages/rfw/example/wasm/lib/main.dart:49]: expireDate, right

Comment (Hixie) [reply packages/rfw/example/wasm/lib/main.dart:49]: https://en.wikipedia.org/wiki/Expiration_date#:~:text=expiry%20date is what I was going for.

**Comment by Hixie** [comment]
Thanks @dnfield!

File: packages/rfw/example/wasm/lib/main.dart (hunk 1 above)
Comment (treeplate) [reply packages/rfw/example/wasm/lib/main.dart:49]: ah, ok

**Pull request merged**